core/src/main/java/contacts/core/Where.kt

```diff
@@ -403,108 +468,256 @@ private class RawContactsTableWhere(whereString: String) : Where<RawContactsFiel
  * table will remain.
  */
 internal fun <T : AbstractDataField> Where<T>.inRawContactsTable(): Where<RawContactsField> =
-    RawContactsTableWhere(
-        toString().replace(Fields.RawContact.Id.columnName, RawContactsFields.Id.columnName)
-    )
+    copyWithNewFieldType { field ->
+        when (field) {
+            Fields.RawContact.Id -> RawContactsFields.Id
+            else -> field // no substitution
+        }
+    }
 
 // endregion
 
 // region Where
 
 /**
- * Each where expression is paired with its mimetype because the contacts Data table uses
- * generic column names (e.g. data1, data2, etc) using the column 'mimetype' to distinguish
- * the type of data in that generic column.
+ * The WHERE clause of a database query made up of a specific type of [Field] ([T]).
  *
- * For example, querying for contacts with name LIKE 'john' AND address LIKE 'colorado';
+ * ## Developer notes
  *
- * ```
- * WHERE (data1 = 'john' AND mimetype = 'vnd.android.cursor.item/name')
- *   AND (data1 = 'colorado' AND mimetype = 'vnd.android.cursor.item/postal-address_v2')
- * ```
+ * The type [T] is not exactly "used" in this class itself. Rather, it is used for adding type
+ * restrictions when constructing instances at compile time. For example, this allows us to create a
+ * function that takes in a Where of GroupsField. The caller of that function can then only provide
+ * a Where composed of one or more GroupsField.
  *
- * This is important because if the mimetypes are not paired with the query;
+ * The type [T] is not enforced within the class itself in order to support mutating functions
+ * such as [inRawContactsTable] and [inContactsTable]. This will allow us to construct a Where<X>
+ * from a Where<Y>.
  *
- * ```
- * WHERE (data1 = 'johnson' AND data1 = 'colorado')
- * ```
+ * ### Binary tree structure
```

Comment:
Oh boy! I happen to have built myself a binary tree 😁 Those interview leet code questions coming in to save the day!!!